### PR TITLE
Handle upgrading channels in TimeoutPacket.

### PR DESCRIPTION
**Pull request by DimitrisJim**
## Description

closes: #4080

### Commit Message / Changelog Entry

```bash
Handle upgrading channels in TimeoutPacket.
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/keeper/timeout.go [hunk 1]
```diff
@@ -83,10 +83,10 @@ func (k Keeper) TimeoutPacket(
 		return types.ErrNoOpMsg
 	}
 
-	if channel.State != types.OPEN {
+	if channel.State != types.OPEN && channel.FlushStatus != types.FLUSHING {
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/timeout.go:86]: rejects all channels that aren't OPEN and that are not currently flushing (which leaves us with `TRYUPGRADE` and `ACKUPGRADE` in a `FLUSHING` state allowed to timeout packets).

File: modules/core/04-channel/keeper/timeout_test.go [hunk 2]
```diff
@@ -115,6 +162,25 @@ func (suite *KeeperTestSuite) TestTimeoutPacket() {
...
+			timeoutHeight := clienttypes.GetSelfHeight(suite.chainB.GetContext())
+
+			sequence, err := path.EndpointA.SendPacket(timeoutHeight, disabledTimeoutTimestamp, ibctesting.MockPacketData)
+			suite.Require().NoError(err)
+			packet = types.NewPacket(ibctesting.MockPacketData, sequence, path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, path.EndpointB.ChannelConfig.PortID, path.EndpointB.ChannelID, timeoutHeight, disabledTimeoutTimestamp)
+			// need to update chainA's client representing chainB to prove missing ack
+			err = path.EndpointA.UpdateClient()
+			suite.Require().NoError(err)
+
+			// Move channel to TRYUPGRADE and to FLUSH_COMPLETE manually.
```
Comment (DimitrisJim) [inline modules/core/04-channel/keeper/timeout_test.go:178]: due to the positioning of the check in `TimeoutPacket` (after we check if `len(commitment) == 0` it doesn't seem feasable we can check the combination `[TRYUPGRADE/FLUSHCOMPLETE]` w/o setting them manually.

**Review by charleenfei — APPROVED** [review]
lgtm for this version! although i think this logic may need to get updated wrt [spec changes](https://github.com/cosmos/ibc/pull/999) 🙈

**Comment by DimitrisJim** [comment]
> although i think this logic may need to get updated wrt

Ah, yeah. It appears most will :smile:

**Comment by DimitrisJim** [comment]
Think it might be worth to close for now and wait for the updates. Will keep issue open to keep track of it :fleur_de_lis:

**Pull request closed without merging**